### ISSUE TITLE
[Clang] [Sema] Returning an initializer list from a lambda/block that returns `void` is accepted

### ISSUE THREAD

**Issue by Sirraide** [llvm/llvm-project#188661]
Clang currently accepts the following (https://godbolt.org/z/ez6ehs7h9):
```c++
void f() {
    (void) [] () -> void { return {}; };
    (void) [] () -> void { return {1}; };
    (void) [] () -> void { return {1, 2}; };
}
```
Every other compiler rejects this. Lambdas and blocks share the same code path for this, that being `Sema::ActOnCapScopeReturnStmt()` in `SemaStmt.cpp`, so blocks suffer from the same issue:
```c++
void f() {
    (void) ^ void { return {}; };
    (void) ^ void { return {1}; };
    (void) ^ void { return {1, 2}; };
}
```

The cause for this is that we specifically don’t issue a diagnostic if `!isa<InitListExpr>(RetValExp)` is true; this was introduced in https://github.com/llvm/llvm-project/commit/74b173e5, and that commit offers no explanation for this change, nor a test case.

As an aside, the diagnostic for lambdas that you get if you write e.g. `return 4` in a lambda that returns `void` (i.e. `err_return_block_has_expr`) refers to the lambda as a ‘void block’ when it should really say ‘void lambda’ or ‘lambda returning void’.

**Comment by Ko496-glitch** [comment]
@Sirraide can i work on this ? You can assign it to me :)

**Comment by Sirraide** [comment]
> re we also fixing the err return from void block to void lambda or lambda returning void ?

You can try doing that; if that causes like 50 tests to fail, then we should move that part into a separate issue since updating that many tests would be rather tedious.

**Comment by Sirraide** [comment]
@Ko496-glitch I took a quick look at the AST, and it seems we end up w/ an `InitListExpr` whose type is `void`, so it’s not surprising that just removing the check doesn’t work; I think we should just disallow _any_ `InitListExpr` here if the return type of the lambda/block is `void`. 

I believe the reason this is happening is because `InitListExpr`s more or less start off typeless and the type is filled in later (because just from looking at an initializer list, it’s impossible to know what type it is supposed to have w/o examining its context).
